gtime/causality/granger_causality.py

```diff
@@ -0,0 +1,213 @@
+
+        y_pred_single = lr_single.predict(data_single)
+        y_pred_joint = lr_joint.predict(data_joint)
+        
+        try:
+            DOF_single = float(data_single.shape[0] - data_single.shape[1]) 
+        except:
+            DOF_single = float(data_single.shape[0] - 1)
+        DOF_joint = float(data_joint.shape[0] - data_joint.shape[1]) -1  
+                
+        lr_single_residues = lr_single._residues
+        lr_joint_residues = lr_joint._residues
+
+        result = []
+        if self.statistics=='all':
+            stat = ['ssr_F', 'ssr_chi2', 'likelihood_chi2', 'zero_F']
+        else:
+            stat = [self.statistics]
+
+        if 'ssr_F' in stat:
+            result_df = pd.DataFrame()
+            f_stat = ((lr_single_residues - lr_joint_residues) / 
+                    lr_joint_residues / self.max_shift * DOF_joint)                            
+            
+            result_df['ssr F-test'] = [f_stat, stats.f.sf(f_stat, self.max_shift, DOF_joint), int(DOF_joint), int(self.max_shift)]
+            result_df.index = ['F-value', 'p-value', 'degrees of freedom', 'number of shifts']
+            result.append(result_df)
+
+        if 'ssr_chi2' in stat:
+            result_df = pd.DataFrame()
+            fgc2 = len(data_single) * (lr_single_residues - lr_joint_residues) / lr_joint_residues
+            result_df['ssr_chi2test'] = [fgc2, stats.chi2.sf(fgc2, self.max_shift), int(DOF_joint), int(self.max_shift)]
+            result_df.index = ['chi2', 'p-value', 'degrees of freedom', 'number of shifts']
+            result.append(result_df)
+        
+        if 'likelihood_chi2' in stat:
+            result_df = pd.DataFrame()
+            lr_single_loglikelihood = loglikelihood(y_pred=y_pred_single, y_true=data[x].loc[data_single.index])
+            lr_joint_loglikelihood = loglikelihood(y_pred=y_pred_joint, y_true=data[x].loc[data_joint.index])
+            
+            lr = -2 * (lr_single_loglikelihood - lr_joint_loglikelihood)
+
+            result_df['likelihood ratio test'] = [lr, stats.chi2.sf(lr, self.max_shift), int(DOF_joint), int(self.max_shift)]
+            result_df.index = ['chi2', 'p-value', 'degrees of freedom','number of shifts']
+            result.append(result_df)
+
+        if 'zero_F' in stat:
+            result_df = pd.DataFrame()
+            r_matrix = np.column_stack((np.zeros((self.max_shift, self.max_shift)),
+                                        np.eye(self.max_shift, self.max_shift),
+                                        np.zeros((self.max_shift, 1))))
+            
+            y_true = data[x].loc[data_joint.index].values 
+            q_matrix = np.zeros(len(r_matrix)) 
+            
+            params = list(lr_joint.coef_)
+            params.append(lr_joint.intercept_)
+            params = np.array(params) 
+            cparams = np.dot(r_matrix, params[:, None])
+            Rbq = cparams - q_matrix
```

Comment:
Same comment with the lowercase, by convention (as I remember):
`variable`
`Classes`
`CONSTANTS`